src/compiler/program.ts

```diff
@@ -1520,7 +1527,297 @@ namespace ts {
+                        if (activeLint.accepted) {
+                            oneAccepted = true;
+                        }
+                        else {
+                            needsReset[i] = true;
+                        }
+                    }
+                }
+                parent = node;
+                if (oneAccepted) {
+                    forEachChild(node, visitNode);
+                }
+                parent = oldParent;
+                for (let i = 0; i < initializedLints.length; i++) {
+                    activeLint = initializedLints[i];
+                    if (!needsReset[i]) {
+                        activeLint.accepted = true;
+                        needsReset[i] = false;
+                    }
+                    if (!activeLint.errored && activeLint.walker.afterVisit) {
+                        try {
+                            activeLint.walker.afterVisit(node, error);
+                        }
+                        catch (e) {
+                            activeLint.errored = true;
+                            diagnostics.push(createExtensionDiagnostic(errorQualifiedName("!!!"), `afterVisit failed with error: ${e}`, /*sourceFile*/undefined, /*start*/undefined, /*length*/undefined, DiagnosticCategory.Error));
+                        }
+                    }
+                }
+            }
+
+            function errorQualifiedName(shortname?: string) {
+                if (!shortname) return activeLint.name;
+                return `${activeLint.name}(${shortname})`;
+            }
+
+            function error(err: string): void;
+            function error(err: string, node: Node): void;
+            function error(err: string, start: number, length: number): void;
+            function error(shortname: string, err: string): void;
+            function error(shortname: string, err: string, span: Node): void;
+            function error(shortname: string, err: string, start: number, length: number): void;
+            function error(level: DiagnosticCategory, err: string): void;
+            function error(level: DiagnosticCategory, err: string, node: Node): void;
+            function error(level: DiagnosticCategory, err: string, start: number, length: number): void;
+            function error(level: DiagnosticCategory, shortname: string, err: string): void;
+            function error(level: DiagnosticCategory, shortname: string, err: string, span: Node): void;
+            function error(level: DiagnosticCategory, shortname: string, err: string, start: number, length: number): void;
+            function error(
+                levelErrOrShortname: DiagnosticCategory | string,
+                errShortnameSpanOrStart?: string | Node | number,
+                errSpanStartOrLength?: string | Node | number,
+                startSpanOrLength?: number | Node,
+                length?: number): void {
+                if (typeof levelErrOrShortname === "undefined") {

```

Comment:
isn't there some kind of Overload package that handles all this dispatch and error checking for you? Admittedly we would be unlikely to add a dependency, but maybe we could use the ideas from them. There's about 150 lines of overhead from this code, so anything a helper could be almost that long and still save code.